### PR TITLE
better error typings for auth0-js

### PR DESCRIPTION
**Pull request by donaldpipowitch**
Please fill in this template.

- [x] Use a meaningful title for the pull request. Include the name of the package modified.
- [ ] Test the change in your own code. (Compile and run.)
- [ ] Add or edit tests to reflect the change. (Run with `npm test`.)
- [ ] Follow the advice from the [readme](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#make-a-pull-request).
- [ ] Avoid [common mistakes](https://github.com/DefinitelyTyped/DefinitelyTyped/blob/master/README.md#common-mistakes).
- [ ] Run `npm run lint package-name` (or `tsc` if no `tslint.json` is present).

Select one of these and delete the others:

If changing an existing definition:
- [x] Provide a URL to documentation or source code which provides context for the suggested changes: See explanation below.
- [ ] Increase the version number in the header if appropriate.
- [ ] If you are making substantial changes, consider adding a `tslint.json` containing `{ "extends": "dtslint/dt.json" }`.

---

This PR tries to fix the errors which can be returned by `auth0-js`. I gathered all information in this issue: https://github.com/auth0/auth0.js/issues/861
I also added `null` as a possible return type for `parseHash` (this was documented here: https://github.com/auth0/auth0.js/issues/766).

I added some inline documentation for some of the error codes which I took from here: https://auth0.com/docs/api-auth/tutorials/silent-authentication#error-response.

### CODE REVIEW

File: types/auth0-js/index.d.ts [hunk 1]
```diff
@@ -482,7 +482,7 @@ export class CrossOriginAuthentication {
     callback(): void;
 }
 
-export type Auth0Callback<T> = (error: null | Auth0Error, result: T) => void;
+export type Auth0Callback<T, E = Auth0Error> = (error: null | E, result: T) => void;
```
Comment (peterblazejewicz) [inline types/auth0-js/index.d.ts]: this requires the 2.3 version (as mentioned in failing CI tests):
https://www.typescriptlang.org/docs/handbook/release-notes/typescript-2-3.html
Can we add the 2.3 requiment pragma(s) as here?
https://github.com/DefinitelyTyped/DefinitelyTyped/pull/29101

**Review by peterblazejewicz — CHANGES_REQUESTED** [review]

File: types/auth0-js/index.d.ts (hunk 1 above)
Comment (donaldpipowitch) [reply types/auth0-js/index.d.ts]: Thank you @peterblazejewicz for the quick review and the useful link!

**Review by peterblazejewicz — COMMENTED** [review]
Was this better solution to remove 2.7 requirement and use that
```ts
"false" as 'false'
```
As I undestand they fixed the issue with Object literals types in 2.7:
https://www.typescriptlang.org/docs/handbook/release-notes/typescript-2-7.html

I know that this fixes the TSC build failure, but maybe that'd be wiser topush dependency version up anyway?

**Comment by EatonZ** [comment]
Hi, would you consider adding an `error_description`? [See here for why.](https://github.com/auth0/auth0.js/issues/866)

It would probably be better for Auth0 to change `checkSession`'s error to `errorDescription` to keep things consistent, though.

**Comment by donaldpipowitch** [comment]
@peterblazejewicz Yeah, I was unsure if I should set 2.7 as the requirement for the whole lib, but I can do that.

@EatonZ Interesting. I updated my original issue about possible errors with another comment. https://github.com/auth0/auth0.js/issues/861 I'd like @luisrudge to answer, if we should actually fix `error_description` inside `auth0-js` itself.

**Review by peterblazejewicz — APPROVED** [review]

**Pull request merged**